src/SAWScript/Value.hs

```diff
@@ -853,13 +868,25 @@ addTrace str val =
 
 -- | Wrap an action with a handler that catches and rethrows user
 -- errors with an extended message.
-addTraceIO :: String -> IO a -> IO a
-addTraceIO str action = X.catchJust p action h
+addTraceIO :: forall a. String -> IO a -> IO a
+addTraceIO str action = X.catches action
+  [ X.Handler handleTopLevel
+  , X.Handler handleTrace
+  , X.Handler handleIO
+  ]
   where
-    -- TODO: Use a custom exception type instead of fail/userError
-    -- init/drop 12 is a hack to remove "user error (" and ")"
-    p e = if IOError.isUserError e then Just (init (drop 12 (show e))) else Nothing
-    h msg = X.throwIO (IOError.userError (str ++ ":\n" ++ msg))
+    rethrow msg = X.throwIO . SS.TraceException $ mconcat [str, ":\n", msg]
+    handleTopLevel :: SS.TopLevelException -> IO a
+    handleTopLevel (SS.TopLevelException _pos msg) = rethrow msg
+    handleTopLevel (SS.JavaException _pos msg) = rethrow msg
+    handleTopLevel (SS.CrucibleSetupException _loc msg) = rethrow msg
+    handleTopLevel (SS.OverrideMatcherException _loc msg) = rethrow msg
+    handleTopLevel (SS.LLVMMethodSpecException _loc msg) = rethrow msg
```

Comment:
So are the locations mostly a placeholder for now? Perhaps we could have a follow-up issue to append them to the messages before the user actually sees them.

Reply:
Yes, for now at least. Appending the locations to the messages here doesn't seem to help much, since the same locations already appear in the backtrace. In the future, it would be nice to fail with better locations (e.g. the location of a specific `crucible_points_to` in the method spec), which would make this more useful.